// @react-awesome-query-builder/ui/esm/config/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
import React from "react";
import * as Widgets from "../components/widgets";
import * as CustomOperators from "../components/operators";
import { CoreConfig } from "@react-awesome-query-builder/core";
var VanillaBooleanWidget = Widgets.VanillaBooleanWidget,
  VanillaTextWidget = Widgets.VanillaTextWidget,
  VanillaTextAreaWidget = Widgets.VanillaTextAreaWidget,
  VanillaDateWidget = Widgets.VanillaDateWidget,
  VanillaTimeWidget = Widgets.VanillaTimeWidget,
  VanillaDateTimeWidget = Widgets.VanillaDateTimeWidget,
  VanillaMultiSelectWidget = Widgets.VanillaMultiSelectWidget,
  VanillaSelectWidget = Widgets.VanillaSelectWidget,
  VanillaNumberWidget = Widgets.VanillaNumberWidget,
  VanillaSliderWidget = Widgets.VanillaSliderWidget,
  VanillaFieldSelect = Widgets.VanillaFieldSelect,
  VanillaConjs = Widgets.VanillaConjs,
  VanillaButton = Widgets.VanillaButton,
  VanillaButtonGroup = Widgets.VanillaButtonGroup,
  VanillaProvider = Widgets.VanillaProvider,
  VanillaValueSources = Widgets.VanillaValueSources,
  vanillaConfirm = Widgets.vanillaConfirm,
  VanillaSwitch = Widgets.VanillaSwitch,
  ValueFieldWidget = Widgets.ValueFieldWidget,
  FuncWidget = Widgets.FuncWidget;
var ProximityOperator = CustomOperators.ProximityOperator;

//----------------------------  conjunctions

var conjunctions = _objectSpread({}, CoreConfig.conjunctions);

//----------------------------  operators

var operators = _objectSpread(_objectSpread({}, CoreConfig.operators), {}, {
  proximity: _objectSpread(_objectSpread({}, CoreConfig.operators.proximity), {}, {
    options: _objectSpread(_objectSpread({}, CoreConfig.operators.proximity.options), {}, {
      factory: function factory(props) {
        return /*#__PURE__*/React.createElement(ProximityOperator, props);
      }
    })
  })
});

//----------------------------  widgets

var widgets = {
  text: _objectSpread(_objectSpread({}, CoreConfig.widgets.text), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(VanillaTextWidget, props);
    }
  }),
  textarea: _objectSpread(_objectSpread({}, CoreConfig.widgets.textarea), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(VanillaTextAreaWidget, props);
    }
  }),
  number: _objectSpread(_objectSpread({}, CoreConfig.widgets.number), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(VanillaNumberWidget, props);
    }
  }),
  slider: _objectSpread(_objectSpread({}, CoreConfig.widgets.slider), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(VanillaSliderWidget, props);
    }
  }),
  select: _objectSpread(_objectSpread({}, CoreConfig.widgets.select), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(VanillaSelectWidget, props);
    }
  }),
  multiselect: _objectSpread(_objectSpread({}, CoreConfig.widgets.multiselect), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(VanillaMultiSelectWidget, props);
    }
  }),
  date: _objectSpread(_objectSpread({}, CoreConfig.widgets.date), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(VanillaDateWidget, props);
    }
  }),
  time: _objectSpread(_objectSpread({}, CoreConfig.widgets.time), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(VanillaTimeWidget, props);
    }
  }),
  datetime: _objectSpread(_objectSpread({}, CoreConfig.widgets.datetime), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(VanillaDateTimeWidget, props);
    }
  }),
  "boolean": _objectSpread(_objectSpread({}, CoreConfig.widgets["boolean"]), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(VanillaBooleanWidget, props);
    }
  }),
  field: _objectSpread(_objectSpread({}, CoreConfig.widgets.field), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(ValueFieldWidget, props);
    },
    customProps: {
      showSearch: true
    }
  }),
  func: _objectSpread(_objectSpread({}, CoreConfig.widgets.func), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/React.createElement(FuncWidget, props);
    },
    customProps: {
      //showSearch: true
    }
  }),
  case_value: _objectSpread(_objectSpread({}, CoreConfig.widgets.case_value), {}, {
    factory: function factory(_ref) {
      var value = _ref.value,
        setValue = _ref.setValue;
      return /*#__PURE__*/React.createElement("input", {
        type: "text",
        value: value || "",
        onChange: function onChange(e) {
          return setValue(e.target.value);
        }
      });
    }
  })
};

//----------------------------  types

var types = _objectSpread(_objectSpread({}, CoreConfig.types), {}, {
  select: _objectSpread(_objectSpread({}, CoreConfig.types.select), {}, {
    widgets: _objectSpread(_objectSpread({}, CoreConfig.types.select.widgets), {}, {
      select: _objectSpread(_objectSpread({}, CoreConfig.types.select.widgets.select), {}, {
        widgetProps: {
          customProps: {
            showSearch: true
          }
        }
      })
    })
  })
});

//----------------------------  settings

var settings = _objectSpread(_objectSpread({}, CoreConfig.settings), {}, {
  renderField: function renderField(props) {
    return /*#__PURE__*/React.createElement(VanillaFieldSelect, props);
  },
  renderOperator: function renderOperator(props) {
    return /*#__PURE__*/React.createElement(VanillaFieldSelect, props);
  },
  renderFunc: function renderFunc(props) {
    return /*#__PURE__*/React.createElement(VanillaFieldSelect, props);
  },
  renderConjs: function renderConjs(props) {
    return /*#__PURE__*/React.createElement(VanillaConjs, props);
  },
  renderSwitch: function renderSwitch(props) {
    return /*#__PURE__*/React.createElement(VanillaSwitch, props);
  },
  renderButton: function renderButton(props) {
    return /*#__PURE__*/React.createElement(VanillaButton, props);
  },
  renderButtonGroup: function renderButtonGroup(props) {
    return /*#__PURE__*/React.createElement(VanillaButtonGroup, props);
  },
  renderProvider: function renderProvider(props) {
    return /*#__PURE__*/React.createElement(VanillaProvider, props);
  },
  renderValueSources: function renderValueSources(props) {
    return /*#__PURE__*/React.createElement(VanillaValueSources, props);
  },
  renderConfirm: vanillaConfirm,
  renderSwitchPrefix: function renderSwitchPrefix() {
    return /*#__PURE__*/React.createElement(React.Fragment, null, "Conditions");
  },
  customFieldSelectProps: {
    showSearch: true
  },
  defaultSliderWidth: "200px",
  defaultSelectWidth: "200px",
  defaultSearchWidth: "100px",
  defaultMaxRows: 5,
  renderSize: "small",
  maxLabelsLength: 100,
  showLock: false,
  showNot: true,
  forceShowConj: false,
  groupActionsPosition: "topRight" // oneOf [topLeft, topCenter, topRight, bottomLeft, bottomCenter, bottomRight]
});

//----------------------------

export default {
  conjunctions: conjunctions,
  operators: operators,
  widgets: widgets,
  types: types,
  settings: settings
};
