// @react-awesome-query-builder/mui/cjs/config/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _widgets = _interopRequireDefault(require("../widgets"));
var _ui = require("@react-awesome-query-builder/ui");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var _Utils$ExportUtils = _ui.Utils.ExportUtils,
  SqlString = _Utils$ExportUtils.SqlString,
  stringifyForDisplay = _Utils$ExportUtils.stringifyForDisplay;
var MuiBooleanWidget = _widgets["default"].MuiBooleanWidget,
  MuiTextWidget = _widgets["default"].MuiTextWidget,
  MuiTextAreaWidget = _widgets["default"].MuiTextAreaWidget,
  MuiDateWidget = _widgets["default"].MuiDateWidget,
  MuiTimeWidget = _widgets["default"].MuiTimeWidget,
  MuiDateTimeWidget = _widgets["default"].MuiDateTimeWidget,
  MuiMultiSelectWidget = _widgets["default"].MuiMultiSelectWidget,
  MuiSelectWidget = _widgets["default"].MuiSelectWidget,
  MuiNumberWidget = _widgets["default"].MuiNumberWidget,
  MuiSliderWidget = _widgets["default"].MuiSliderWidget,
  MuiRangeWidget = _widgets["default"].MuiRangeWidget,
  MuiAutocompleteWidget = _widgets["default"].MuiAutocompleteWidget,
  MuiFieldSelect = _widgets["default"].MuiFieldSelect,
  MuiFieldAutocomplete = _widgets["default"].MuiFieldAutocomplete,
  MuiConjs = _widgets["default"].MuiConjs,
  MuiSwitch = _widgets["default"].MuiSwitch,
  MuiButton = _widgets["default"].MuiButton,
  MuiButtonGroup = _widgets["default"].MuiButtonGroup,
  MuiValueSources = _widgets["default"].MuiValueSources,
  MuiProvider = _widgets["default"].MuiProvider,
  MuiConfirm = _widgets["default"].MuiConfirm,
  MuiUseConfirm = _widgets["default"].MuiUseConfirm;
var settings = _objectSpread(_objectSpread({}, _ui.BasicConfig.settings), {}, {
  renderField: function renderField(props) {
    var _props$customProps;
    return props !== null && props !== void 0 && (_props$customProps = props.customProps) !== null && _props$customProps !== void 0 && _props$customProps.showSearch ? /*#__PURE__*/_react["default"].createElement(MuiFieldAutocomplete, props) : /*#__PURE__*/_react["default"].createElement(MuiFieldSelect, props);
  },
  renderOperator: function renderOperator(props) {
    return /*#__PURE__*/_react["default"].createElement(MuiFieldSelect, props);
  },
  renderFunc: function renderFunc(props) {
    return /*#__PURE__*/_react["default"].createElement(MuiFieldSelect, props);
  },
  renderConjs: function renderConjs(props) {
    return /*#__PURE__*/_react["default"].createElement(MuiConjs, props);
  },
  renderSwitch: function renderSwitch(props) {
    return /*#__PURE__*/_react["default"].createElement(MuiSwitch, props);
  },
  renderButton: function renderButton(props) {
    return /*#__PURE__*/_react["default"].createElement(MuiButton, props);
  },
  renderButtonGroup: function renderButtonGroup(props) {
    return /*#__PURE__*/_react["default"].createElement(MuiButtonGroup, props);
  },
  renderValueSources: function renderValueSources(props) {
    return /*#__PURE__*/_react["default"].createElement(MuiValueSources, props);
  },
  renderProvider: function renderProvider(props) {
    return /*#__PURE__*/_react["default"].createElement(MuiProvider, props);
  },
  renderConfirm: MuiConfirm,
  useConfirm: MuiUseConfirm
});
var widgets = _objectSpread(_objectSpread({}, _ui.BasicConfig.widgets), {}, {
  text: _objectSpread(_objectSpread({}, _ui.BasicConfig.widgets.text), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(MuiTextWidget, props);
    }
  }),
  textarea: _objectSpread(_objectSpread({}, _ui.BasicConfig.widgets.textarea), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(MuiTextAreaWidget, props);
    }
  }),
  number: _objectSpread(_objectSpread({}, _ui.BasicConfig.widgets.number), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(MuiNumberWidget, props);
    }
  }),
  multiselect: _objectSpread(_objectSpread({}, _ui.BasicConfig.widgets.multiselect), {}, {
    factory: function factory(props) {
      return props.asyncFetch || props.showSearch ? /*#__PURE__*/_react["default"].createElement(MuiAutocompleteWidget, (0, _extends2["default"])({
        multiple: true
      }, props)) : /*#__PURE__*/_react["default"].createElement(MuiMultiSelectWidget, props);
    }
  }),
  select: _objectSpread(_objectSpread({}, _ui.BasicConfig.widgets.select), {}, {
    factory: function factory(props) {
      return props.asyncFetch || props.showSearch ? /*#__PURE__*/_react["default"].createElement(MuiAutocompleteWidget, props) : /*#__PURE__*/_react["default"].createElement(MuiSelectWidget, props);
    }
  }),
  slider: _objectSpread(_objectSpread({}, _ui.BasicConfig.widgets.slider), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(MuiSliderWidget, props);
    }
  }),
  "boolean": _objectSpread(_objectSpread({}, _ui.BasicConfig.widgets["boolean"]), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(MuiBooleanWidget, props);
    }
  }),
  date: _objectSpread(_objectSpread({}, _ui.BasicConfig.widgets.date), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(MuiDateWidget, props);
    }
  }),
  time: _objectSpread(_objectSpread({}, _ui.BasicConfig.widgets.time), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(MuiTimeWidget, props);
    }
  }),
  datetime: _objectSpread(_objectSpread({}, _ui.BasicConfig.widgets.datetime), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(MuiDateTimeWidget, props);
    }
  }),
  rangeslider: {
    type: "number",
    jsType: "number",
    valueSrc: "value",
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(MuiRangeWidget, props);
    },
    valueLabel: "Range",
    valuePlaceholder: "Select range",
    valueLabels: [{
      label: "Number from",
      placeholder: "Enter number from"
    }, {
      label: "Number to",
      placeholder: "Enter number to"
    }],
    formatValue: function formatValue(val, fieldDef, wgtDef, isForDisplay) {
      return isForDisplay ? stringifyForDisplay(val) : JSON.stringify(val);
    },
    sqlFormatValue: function sqlFormatValue(val, fieldDef, wgtDef, op, opDef) {
      return SqlString.escape(val);
    },
    singleWidget: "slider",
    toJS: function toJS(val, fieldSettings) {
      return val;
    }
  }
});
var types = _objectSpread(_objectSpread({}, _ui.BasicConfig.types), {}, {
  number: _objectSpread(_objectSpread({}, _ui.BasicConfig.types.number), {}, {
    widgets: _objectSpread(_objectSpread({}, _ui.BasicConfig.types.number.widgets), {}, {
      rangeslider: {
        opProps: {
          between: {
            isSpecialRange: true
          },
          not_between: {
            isSpecialRange: true
          }
        },
        operators: ["between", "not_between", "is_empty", "is_not_empty"]
      }
    })
  })
});
var _default = _objectSpread(_objectSpread({}, _ui.BasicConfig), {}, {
  types: types,
  widgets: widgets,
  settings: settings
});
exports["default"] = _default;
