// @react-awesome-query-builder/bootstrap/esm/widgets/value/BootstrapSlider.js
import React from "react";
import { Input } from "reactstrap";
export default (function (props) {
  var value = props.value,
    setValue = props.setValue,
    config = props.config,
    readonly = props.readonly,
    min = props.min,
    max = props.max,
    step = props.step,
    placeholder = props.placeholder;
  var onChange = function onChange(e) {
    var val = e.target.value;
    if (val === "" || val === null) val = undefined;else val = Number(val);
    setValue(val);
  };
  var stylesWrapper = {
    display: "inline-flex",
    flexWrap: "wrap"
  };
  var stylesInputWrapper = {
    marginLeft: "5px",
    display: "inline-flex",
    width: "auto"
  };
  var stylesSliderWrapper = {
    marginLeft: "5px",
    display: "inline-flex",
    width: "auto",
    minWidth: "150px"
  };
  var numberValue = value == undefined ? "" : value;
  return /*#__PURE__*/React.createElement("div", {
    style: stylesWrapper
  }, /*#__PURE__*/React.createElement(Input, {
    key: "number",
    bsSize: "sm",
    style: stylesInputWrapper,
    type: "number",
    value: numberValue,
    placeholder: placeholder,
    disabled: readonly,
    min: min,
    max: max,
    step: step,
    onChange: onChange
  }), /*#__PURE__*/React.createElement(Input, {
    key: "range",
    bsSize: "sm",
    style: stylesSliderWrapper,
    type: "range",
    value: numberValue,
    disabled: readonly,
    min: min,
    max: max,
    step: step,
    onChange: onChange
  }));
});
