// @react-awesome-query-builder/bootstrap/esm/widgets/core/BootstrapFieldSelect.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import React, { useState } from "react";
import { Dropdown, DropdownMenu, DropdownToggle, DropdownItem } from "reactstrap";
export default (function (_ref) {
  var items = _ref.items,
    setField = _ref.setField,
    selectedKey = _ref.selectedKey,
    readonly = _ref.readonly,
    placeholder = _ref.placeholder;
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    isOpen = _useState2[0],
    setIsOpen = _useState2[1];
  var stylesDropdownWrapper = {
    lineHeight: "105%",
    minHeight: "1.7rem",
    paddingBottom: "0.45rem"
  };
  var stylesDropdownMenuWrapper = {
    overflowY: "auto",
    maxHeight: "400px"
  };
  var onChange = function onChange(e) {
    if (e.target.value === undefined) return;
    setField(e.target.value);
  };
  var renderOptions = function renderOptions(fields) {
    var isGroupItem = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
    return Object.keys(fields).map(function (fieldKey) {
      var field = fields[fieldKey];
      var items = field.items,
        path = field.path,
        label = field.label,
        disabled = field.disabled;
      if (items) {
        return /*#__PURE__*/React.createElement("div", {
          key: "dropdown-itemGroup-".concat(path)
        }, /*#__PURE__*/React.createElement(DropdownItem, {
          header: true,
          key: "".concat(path, "-header"),
          onClick: onChange,
          value: path
        }, label), renderOptions(items, true));
      } else {
        return /*#__PURE__*/React.createElement(DropdownItem, {
          disabled: disabled,
          key: path,
          onClick: onChange,
          value: path,
          className: isGroupItem ? "px-4" : undefined,
          active: selectedKey == path
        }, label);
      }
    });
  };
  var hasValue = selectedKey != null;
  var renderValue = function renderValue(selectedValue) {
    if (!readonly && !selectedValue) return placeholder;
    var findLabel = function findLabel(fields) {
      return fields.map(function (field) {
        if (!field.items) return field.path === selectedValue ? field.label : null;
        return findLabel(field.items);
      });
    };
    return findLabel(items).filter(function (v) {
      if (Array.isArray(v)) {
        return v.some(function (value) {
          return value !== null;
        });
      } else {
        return v !== null;
      }
    }).pop();
  };
  return /*#__PURE__*/React.createElement(Dropdown, {
    isOpen: isOpen,
    onClick: function onClick() {
      return !isOpen ? setIsOpen(true) : setIsOpen(false);
    },
    disabled: readonly,
    toggle: function toggle() {
      return setIsOpen(!isOpen);
    }
  }, /*#__PURE__*/React.createElement(DropdownToggle, {
    tag: "button",
    className: "form-select",
    style: stylesDropdownWrapper,
    color: "transparent"
  }, hasValue ? renderValue(selectedKey) : /*#__PURE__*/React.createElement("span", null, "\xA0")), /*#__PURE__*/React.createElement(DropdownMenu, {
    container: "body",
    style: stylesDropdownMenuWrapper
  }, !hasValue && /*#__PURE__*/React.createElement(DropdownItem, {
    key: "body",
    disabled: true,
    value: ""
  }), renderOptions(items)));
});
