// @react-awesome-query-builder/ui/cjs/config/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var Widgets = _interopRequireWildcard(require("../components/widgets"));
var CustomOperators = _interopRequireWildcard(require("../components/operators"));
var _core = require("@react-awesome-query-builder/core");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var VanillaBooleanWidget = Widgets.VanillaBooleanWidget,
  VanillaTextWidget = Widgets.VanillaTextWidget,
  VanillaTextAreaWidget = Widgets.VanillaTextAreaWidget,
  VanillaDateWidget = Widgets.VanillaDateWidget,
  VanillaTimeWidget = Widgets.VanillaTimeWidget,
  VanillaDateTimeWidget = Widgets.VanillaDateTimeWidget,
  VanillaMultiSelectWidget = Widgets.VanillaMultiSelectWidget,
  VanillaSelectWidget = Widgets.VanillaSelectWidget,
  VanillaNumberWidget = Widgets.VanillaNumberWidget,
  VanillaSliderWidget = Widgets.VanillaSliderWidget,
  VanillaFieldSelect = Widgets.VanillaFieldSelect,
  VanillaConjs = Widgets.VanillaConjs,
  VanillaButton = Widgets.VanillaButton,
  VanillaButtonGroup = Widgets.VanillaButtonGroup,
  VanillaProvider = Widgets.VanillaProvider,
  VanillaValueSources = Widgets.VanillaValueSources,
  vanillaConfirm = Widgets.vanillaConfirm,
  VanillaSwitch = Widgets.VanillaSwitch,
  ValueFieldWidget = Widgets.ValueFieldWidget,
  FuncWidget = Widgets.FuncWidget;
var ProximityOperator = CustomOperators.ProximityOperator;

//----------------------------  conjunctions

var conjunctions = _objectSpread({}, _core.CoreConfig.conjunctions);

//----------------------------  operators

var operators = _objectSpread(_objectSpread({}, _core.CoreConfig.operators), {}, {
  proximity: _objectSpread(_objectSpread({}, _core.CoreConfig.operators.proximity), {}, {
    options: _objectSpread(_objectSpread({}, _core.CoreConfig.operators.proximity.options), {}, {
      factory: function factory(props) {
        return /*#__PURE__*/_react["default"].createElement(ProximityOperator, props);
      }
    })
  })
});

//----------------------------  widgets

var widgets = {
  text: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.text), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(VanillaTextWidget, props);
    }
  }),
  textarea: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.textarea), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(VanillaTextAreaWidget, props);
    }
  }),
  number: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.number), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(VanillaNumberWidget, props);
    }
  }),
  slider: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.slider), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(VanillaSliderWidget, props);
    }
  }),
  select: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.select), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(VanillaSelectWidget, props);
    }
  }),
  multiselect: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.multiselect), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(VanillaMultiSelectWidget, props);
    }
  }),
  date: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.date), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(VanillaDateWidget, props);
    }
  }),
  time: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.time), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(VanillaTimeWidget, props);
    }
  }),
  datetime: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.datetime), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(VanillaDateTimeWidget, props);
    }
  }),
  "boolean": _objectSpread(_objectSpread({}, _core.CoreConfig.widgets["boolean"]), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(VanillaBooleanWidget, props);
    }
  }),
  field: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.field), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(ValueFieldWidget, props);
    },
    customProps: {
      showSearch: true
    }
  }),
  func: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.func), {}, {
    factory: function factory(props) {
      return /*#__PURE__*/_react["default"].createElement(FuncWidget, props);
    },
    customProps: {
      //showSearch: true
    }
  }),
  case_value: _objectSpread(_objectSpread({}, _core.CoreConfig.widgets.case_value), {}, {
    factory: function factory(_ref) {
      var value = _ref.value,
        setValue = _ref.setValue;
      return /*#__PURE__*/_react["default"].createElement("input", {
        type: "text",
        value: value || "",
        onChange: function onChange(e) {
          return setValue(e.target.value);
        }
      });
    }
  })
};

//----------------------------  types

var types = _objectSpread(_objectSpread({}, _core.CoreConfig.types), {}, {
  select: _objectSpread(_objectSpread({}, _core.CoreConfig.types.select), {}, {
    widgets: _objectSpread(_objectSpread({}, _core.CoreConfig.types.select.widgets), {}, {
      select: _objectSpread(_objectSpread({}, _core.CoreConfig.types.select.widgets.select), {}, {
        widgetProps: {
          customProps: {
            showSearch: true
          }
        }
      })
    })
  })
});

//----------------------------  settings

var settings = _objectSpread(_objectSpread({}, _core.CoreConfig.settings), {}, {
  renderField: function renderField(props) {
    return /*#__PURE__*/_react["default"].createElement(VanillaFieldSelect, props);
  },
  renderOperator: function renderOperator(props) {
    return /*#__PURE__*/_react["default"].createElement(VanillaFieldSelect, props);
  },
  renderFunc: function renderFunc(props) {
    return /*#__PURE__*/_react["default"].createElement(VanillaFieldSelect, props);
  },
  renderConjs: function renderConjs(props) {
    return /*#__PURE__*/_react["default"].createElement(VanillaConjs, props);
  },
  renderSwitch: function renderSwitch(props) {
    return /*#__PURE__*/_react["default"].createElement(VanillaSwitch, props);
  },
  renderButton: function renderButton(props) {
    return /*#__PURE__*/_react["default"].createElement(VanillaButton, props);
  },
  renderButtonGroup: function renderButtonGroup(props) {
    return /*#__PURE__*/_react["default"].createElement(VanillaButtonGroup, props);
  },
  renderProvider: function renderProvider(props) {
    return /*#__PURE__*/_react["default"].createElement(VanillaProvider, props);
  },
  renderValueSources: function renderValueSources(props) {
    return /*#__PURE__*/_react["default"].createElement(VanillaValueSources, props);
  },
  renderConfirm: vanillaConfirm,
  renderSwitchPrefix: function renderSwitchPrefix() {
    return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, "Conditions");
  },
  customFieldSelectProps: {
    showSearch: true
  },
  defaultSliderWidth: "200px",
  defaultSelectWidth: "200px",
  defaultSearchWidth: "100px",
  defaultMaxRows: 5,
  renderSize: "small",
  maxLabelsLength: 100,
  showLock: false,
  showNot: true,
  forceShowConj: false,
  groupActionsPosition: "topRight" // oneOf [topLeft, topCenter, topRight, bottomLeft, bottomCenter, bottomRight]
});

//----------------------------
var _default = {
  conjunctions: conjunctions,
  operators: operators,
  widgets: widgets,
  types: types,
  settings: settings
};
exports["default"] = _default;
